### PR TITLE
Update CI images and workflow actions

### PR DESCRIPTION
**Pull request by sethrj**
Lots of our github actions are showing warnings [because github has deprecated Node 16](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/), so we need to update our actions.

I've moved all of the no-HIP rocm builds to native github runners using spack and enabled ccache, so hopefully the pipeline time will decrease.

The new actions do *not* work on the versions of centos/ubuntu used by our GPU docker images, so I'm going to set up new docker containers soon (and finally expand geant4 testing to get the whole 10.5–11.2 range).

### CODE REVIEW

File: .github/workflows/build-docker.yml [hunk 1]
```diff
@@ -11,7 +11,13 @@ jobs:
   # TODO: this currently includes non-GPU builds as well
   # and won't actually run on device
   gpu:
-    name: gpu
+    name: >-
+      ${{format('{0}-{1}-{1}{2}{3}',
```
Comment (pcanal) [inline .github/workflows/build-docker.yml]: `{1}` is repeated and the last of the 5 arguments seems ignored, is that intentional?

File: .github/workflows/build-fast.yml [hunk 2]
```diff
@@ -52,30 +54,22 @@ jobs:
...
-          restore-keys: ccache-${{matrix.runner}}-${{matrix.compiler}}-${{matrix.version}}
+          key: ccache-${{env.GHA_JOB_NAME}}-${{github.run_id}}
+          restore-keys: |
+            ccache-${{env.GHA_JOB_NAME}}
       - name: Zero ccache stats
         run: |
           ccache -z
       - name: Configure Celeritas
         run: |
-          mkdir build && cd build
```
Comment (pcanal) [inline .github/workflows/build-fast.yml:65]: Why is the creation of the build directory no longer needed?

File: .github/workflows/build-docker.yml [hunk 3]
```diff
@@ -25,38 +31,19 @@ jobs:
           - buildtype: 'debug'
             image: 'centos-rocm' # Debug builds don't work with HIP
         include:
-          - special: 'asan'
```
Comment (pcanal) [inline .github/workflows/build-docker.yml:28]: I don't see the `asan` part of the matrix moved anywhere.  Is this for later or did I miss it?

File: .github/workflows/build-fast.yml [hunk 4]
```diff
@@ -26,16 +26,18 @@ jobs:
...
+      GHA_JOB_NAME: "${{matrix.runner}}-${{matrix.compiler}}-${{matrix.version}}"
       CCACHE_DIR: "${{github.workspace}}/.ccache"
       CCACHE_MAXSIZE: "10G"
+      CMAKE_PRESET: fast
       CC: ${{matrix.compiler}}-${{matrix.version}}
       CXX: ${{matrix.compiler == 'gcc' && 'g++' || 'clang++'}}-${{matrix.version}}
+    runs-on: >-
+      ${{  matrix.runner == 'focal' && 'ubuntu-20.04'
+        || matrix.runner == 'jammy' && 'ubuntu-22.04'
+        || null
```
Comment (pcanal) [inline .github/workflows/build-fast.yml]: (pre-existing).  Can you remind me what is the purpose of that `|| false`

File: .github/workflows/build-fast.yml (hunk 2 above)
Comment (sethrj) [reply .github/workflows/build-fast.yml:65]: It's handled by cmake presets now

File: .github/workflows/build-docker.yml (hunk 1 above)
Comment (sethrj) [reply .github/workflows/build-docker.yml]: Nope! Thanks for the eagle eyed catch.

File: .github/workflows/build-docker.yml (hunk 3 above)
Comment (sethrj) [reply .github/workflows/build-docker.yml:28]: Ooh also good catch, I meant to move this to `build-spack` but apparently forgot.

File: .github/workflows/build-fast.yml (hunk 4 above)
Comment (sethrj) [reply .github/workflows/build-fast.yml]: I just put it there to cause an understandable error if the `runner` isn't one of the preselected types?

File: .github/workflows/build-fast.yml (hunk 2 above)
Comment (pcanal) [reply .github/workflows/build-fast.yml:65]: So preset as a feature that create the directory requested when they are not there yet?

File: .github/workflows/build-fast.yml (hunk 4 above)
Comment (pcanal) [reply .github/workflows/build-fast.yml]: I am not sure how that works out to changing the error .... with or without the `|| null` the expression is logically the same, isn't it?

File: .github/workflows/build-fast.yml (hunk 2 above)
Comment (sethrj) [reply .github/workflows/build-fast.yml:65]: Correct:
```json
{
  "configurePresets": [
    {
      "binaryDir": "${sourceDir}/build",
```

File: .github/workflows/build-fast.yml (hunk 4 above)
Comment (sethrj) [reply .github/workflows/build-fast.yml]: Um. I guess you're right, the expression would evaulate to `false`? I'd been thinking of this more as a C-style ternary operator. I'll delete the unnecessary last line.

Comment (pcanal) [reply .github/workflows/build-fast.yml]: >  the expression would evaulate to false?

Yes, it is explicitly send in the github action doc on expression that `null` turns to `false`.

**Review by pcanal — APPROVED** [review]
LGTM.  Thanks.

**Pull request merged**